### PR TITLE
Feature: Use modern CMake

### PR DESCRIPTION
**Pull request by zaufi**
- Update the `readline` finder module to support components, imported targets, and the version check;
- modernize the `CMakeLists.txt`s;
- use `#include "..."` for *project-local* headers;
- leave some more TODOs for maintainers.

### CODE REVIEW

**Review by mutanabbi — COMMENTED** [review]
I don't know. I usually use different directory for this purpose. Does it make any sense to do out of source build inside a source directory? It even sound like kinda joke.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,21 +1,31 @@
...
-)
-enable_testing()
-#set(GTEST_ROOT "/usr/local/lib/")
+    # NOTE Keep the list of components sorted
+    thread
+  )
+
+find_package(Readline 7.0 REQUIRED COMPONENTS readline)
 find_package(GTest)
-find_package(Threads)
```
Comment (mutanabbi) [inline CMakeLists.txt:18]: What about Threads?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,21 +1,31 @@
-cmake_minimum_required(VERSION 3.3)
+# ALERT Always use the latest available release!
+cmake_minimum_required(VERSION 3.14)
```
Comment (mutanabbi) [inline CMakeLists.txt:2]: Why? I prefer to use minimal functional version. If you don't use features from last version, why do you force to use it?

**Review by mutanabbi — COMMENTED** [review]
There are few questions

**Comment by zaufi** [comment]
The point to do _out-of-source_ build is to avoid source directory pollution, and the 2nd is quickly removing the build artifacts (w/ a single command). Having the `build/` directory inside the source tree is just a matter of convenience -- you don't need to go some other place (really out of the source tree).

File: CMakeLists.txt (hunk 1 above)
Comment (zaufi) [reply CMakeLists.txt:18]: The key point is to find only direct dependencies -- the code doesn't `#include <pthread.h>` explicitly.

File: CMakeLists.txt (hunk 2 above)
Comment (zaufi) [reply CMakeLists.txt:2]: Cuz, 'cept the features latest releases have bug fixes, policy sets and probably some changes in defaults (including behavior).

Comment (zaufi) [reply CMakeLists.txt:2]: I didn't know what version do you guys have, so set the latest I have in my broken system. The current one is 3.15.2.

File: CMakeLists.txt (hunk 1 above)
Comment (zaufi) [reply CMakeLists.txt:18]: BTW, need to check if compiler CLI got the `-pthread` option anyway.

Comment (zaufi) [reply CMakeLists.txt:18]: It should come from build requirements of `boost::thread`...

Comment (mutanabbi) [reply CMakeLists.txt:18]: I'm definitely going to use std::threads, but right now there is no MT in code.

**Pull request merged**